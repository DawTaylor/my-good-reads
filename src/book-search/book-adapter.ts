import { Book } from "../book-card/BookCard";

export interface BookListResponse {
  items: BookResponse[],
  totalItems: number
}

interface BookResponse {
  id: string,
  volumeInfo: BookVolumeResponse
}

interface BookVolumeResponse {
  title: string,
  authors: string[],
  publisher: string,
  publishedDate: string,
  description: string,
  imageLinks:  BookVolumeImageLinksResponse
}

interface BookVolumeImageLinksResponse {
  thumbnail: string
}

const bookAdapter = (book: BookResponse): Book | undefined => {
  if (
      !book.id ||
      !book.volumeInfo ||
      !book.volumeInfo.title ||
      !book.volumeInfo.authors ||
      !book.volumeInfo.publisher ||
      !book.volumeInfo.publishedDate ||
      !book.volumeInfo.description ||
      !book.volumeInfo.imageLinks ||
      !book.volumeInfo.imageLinks.thumbnail
  ) return

  return {
      id: book.id,
      coverUrl: book.volumeInfo.imageLinks.thumbnail,
      title: book.volumeInfo.title,
      authors: book.volumeInfo.authors,
      publisher: book.volumeInfo.publisher,
      publishedDate: new Date(book.volumeInfo.publishedDate),
      description: book.volumeInfo.description
  }
}

function notUndefined<T>(x: T | undefined): x is T {
  return x !== undefined;
}

export const bookListAdapter = (bookResponse: BookListResponse): Book[] => {
  if(!bookResponse.totalItems) return []
  return bookResponse.items.map(bookAdapter).filter(notUndefined)
}