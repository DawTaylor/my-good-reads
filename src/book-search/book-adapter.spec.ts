import { bookListAdapter, BookListResponse } from './book-adapter'

import json from '../../sampleJSONResponse.json'

describe('Test book list response adapter', () => {
  test('should return all books in correct format', () => {
    const adaptedBooks = bookListAdapter(json as BookListResponse)

    expect(adaptedBooks).toMatchSnapshot()
  })
})