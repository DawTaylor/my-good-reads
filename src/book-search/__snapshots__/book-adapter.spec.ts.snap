// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test book list response adapter should return all books in correct format 1`] = `
Array [
  Object {
    "authors": Array [
      "David Sawyer McFarland",
    ],
    "coverUrl": "http://books.google.com/books/content?id=ptiYBAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "JavaScript lets you supercharge your HTML with animation, interactivity, and visual effects—but many web designers find the language hard to learn. This easy-to-read guide not only covers JavaScript basics, but also shows you how to save time and effort with the jQuery and jQuery UI libraries of prewritten JavaScript code. You’ll build web pages that feel and act like desktop programs—with little or no programming. The important stuff you need to know: Pull back the curtain on JavaScript. Learn how to build a basic program with this language. Get up to speed on jQuery. Quickly assemble JavaScript programs that work well on multiple web browsers. Transform your user interface. Learn jQuery UI, the JavaScript library for interface features like design themes and controls. Make your pages interactive. Create JavaScript events that react to visitor actions. Use animations and effects. Build drop-down navigation menus, pop-ups, automated slideshows, and more. Collect data with web forms. Create easy-to-use forms that ensure more accurate visitor responses. Practice with living examples. Get step-by-step tutorials for web projects you can build yourself.",
    "id": "ptiYBAAAQBAJ",
    "publishedDate": 2014-09-18T00:00:00.000Z,
    "publisher": "\\"O'Reilly Media, Inc.\\"",
    "title": "JavaScript & jQuery: The Missing Manual",
  },
  Object {
    "authors": Array [
      "Marijn Haverbeke",
    ],
    "coverUrl": "http://books.google.com/books/content?id=UAYvDwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "JavaScript is at the heart of almost every modern Web application, whether it's Google Apps, Twitter, or the newest browser-based game. Though it's simple for beginners to pick up and play with, JavaScript is not a toy—it's a flexible and complex language that can be used to build full-scale applications. Eloquent JavaScript dives into this flourishing language and teaches you to write code that's beautiful and effective. By immersing you in example code and encouraging experimentation right from the start, the author quickly gives you the tools you need to build your own programs. As you follow along with examples like an artificial life simulation and a version of the classic game Sokoban, you'll learn to: –Understand the essential elements of programming: syntax, control, and data –Use object-oriented and functional programming techniques to organize and clarify your programs –Script the browser and make basic Web applications –Work with tools like regular expressions and XMLHttpRequest objects And since programming is an art that's best learned by doing, all example code is available online in an interactive sandbox for you to experiment with. With Eloquent JavaScript as your guide, you can tweak, expand, and modify the author's code, or throw it away and build your own creations from scratch. Before you know it, you'll be fluent in the language of the Web.",
    "id": "UAYvDwAAQBAJ",
    "publishedDate": 2011-01-15T00:00:00.000Z,
    "publisher": "No Starch Press",
    "title": "Eloquent JavaScript",
  },
  Object {
    "authors": Array [
      "Douglas Crockford",
    ],
    "coverUrl": "http://books.google.com/books/content?id=PXa2bby0oQ0C&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "Most programming languages contain good and bad parts, but JavaScript has more than its share of the bad, having been developed and released in a hurry before it could be refined. This authoritative book scrapes away these bad features to reveal a subset of JavaScript that's more reliable, readable, and maintainable than the language as a whole—a subset you can use to create truly extensible and efficient code. Considered the JavaScript expert by many people in the development community, author Douglas Crockford identifies the abundance of good ideas that make JavaScript an outstanding object-oriented programming language-ideas such as functions, loose typing, dynamic objects, and an expressive object literal notation. Unfortunately, these good ideas are mixed in with bad and downright awful ideas, like a programming model based on global variables. When Java applets failed, JavaScript became the language of the Web by default, making its popularity almost completely independent of its qualities as a programming language. In JavaScript: The Good Parts, Crockford finally digs through the steaming pile of good intentions and blunders to give you a detailed look at all the genuinely elegant parts of JavaScript, including: Syntax Objects Functions Inheritance Arrays Regular expressions Methods Style Beautiful features The real beauty? As you move ahead with the subset of JavaScript that this book presents, you'll also sidestep the need to unlearn all the bad parts. Of course, if you want to find out more about the bad parts and how to use them badly, simply consult any other JavaScript book. With JavaScript: The Good Parts, you'll discover a beautiful, elegant, lightweight and highly expressive language that lets you create effective code, whether you're managing object libraries or just trying to get Ajax to run fast. If you develop sites or applications for the Web, this book is an absolute must.",
    "id": "PXa2bby0oQ0C",
    "publishedDate": 2008-05-08T00:00:00.000Z,
    "publisher": "\\"O'Reilly Media, Inc.\\"",
    "title": "JavaScript: The Good Parts",
  },
  Object {
    "authors": Array [
      "David Flanagan",
    ],
    "coverUrl": "http://books.google.com/books/content?id=2weL0iAfrEMC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "This Fifth Edition is completely revised and expanded to cover JavaScript as it is used in today's Web 2.0 applications. This book is both an example-driven programmer's guide and a keep-on-your-desk reference, with new chapters that explain everything you need to know to get the most out of JavaScript, including: Scripted HTTP and Ajax XML processing Client-side graphics using the canvas tag Namespaces in JavaScript--essential when writing complex programs Classes, closures, persistence, Flash, and JavaScript embedded in Java applications Part I explains the core JavaScript language in detail. If you are new to JavaScript, it will teach you the language. If you are already a JavaScript programmer, Part I will sharpen your skills and deepen your understanding of the language. Part II explains the scripting environment provided by web browsers, with a focus on DOM scripting with unobtrusive JavaScript. The broad and deep coverage of client-side JavaScript is illustrated with many sophisticated examples that demonstrate how to: Generate a table of contents for an HTML document Display DHTML animations Automate form validation Draw dynamic pie charts Make HTML elements draggable Define keyboard shortcuts for web applications Create Ajax-enabled tool tips Use XPath and XSLT on XML documents loaded with Ajax And much more Part III is a complete reference for core JavaScript. It documents every class, object, constructor, method, function, property, and constant defined by JavaScript 1.5 and ECMAScript Version 3. Part IV is a reference for client-side JavaScript, covering legacy web browser APIs, the standard Level 2 DOM API, and emerging standards such as the XMLHttpRequest object and the canvas tag. More than 300,000 JavaScript programmers around the world have madethis their indispensable reference book for building JavaScript applications. \\"A must-have reference for expert JavaScript programmers...well-organized and detailed.\\" -- Brendan Eich, creator of JavaScript",
    "id": "2weL0iAfrEMC",
    "publishedDate": 2006-08-17T00:00:00.000Z,
    "publisher": "\\"O'Reilly Media, Inc.\\"",
    "title": "JavaScript: The Definitive Guide",
  },
  Object {
    "authors": Array [
      "thau!",
    ],
    "coverUrl": "http://books.google.com/books/content?id=MrHaC5HAva8C&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "Explains how to use the programming language to add interactivity and animation to Web sites, covering image swaps, functions, frames, cookies, alarms, frames, shopping carts, and Ajax.",
    "id": "MrHaC5HAva8C",
    "publishedDate": 2007-01-01T00:00:00.000Z,
    "publisher": "No Starch Press",
    "title": "The Book of JavaScript, 2nd Edition",
  },
  Object {
    "authors": Array [
      "David Flanagan",
    ],
    "coverUrl": "http://books.google.com/books/content?id=vJGlu9t9LNYC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "A guide for experienced programmers demonstrates the core JavaScript language, offers examples of common tasks, and contains an extensive reference to JavaScript commands, objects, methods, and properties.",
    "id": "vJGlu9t9LNYC",
    "publishedDate": 2002-01-01T00:00:00.000Z,
    "publisher": "\\"O'Reilly Media, Inc.\\"",
    "title": "JavaScript",
  },
  Object {
    "authors": Array [
      "Peter Kahrel",
    ],
    "coverUrl": "http://books.google.com/books/content?id=OW_6BWEQAo4C&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "This Short Cut covers only InDesign CS2. For coverage of InDesign CS3 and CS4, see the version at http://oreilly.com/catalog/9780596802523​/. InDesign provides a powerful set of tools for producing beautiful documents. While you can certainly do all your work by hand through InDesign's graphical interface, there are many times when it's much easier to write a script. Once you've automated a task, you can run it over the whole document, ensuring consistency, or just when you need it, simplifying and speeding your layout process. All it takes is a bit of JavaScript knowledge and a willingness to explore InDesign's programming features. (This version covers only InDesign CS2. For coverage of InDesign CS3 and CS4, see the version at http://oreilly.com/catalog/9780596802523​/.)",
    "id": "OW_6BWEQAo4C",
    "publishedDate": 2006-08-25T00:00:00.000Z,
    "publisher": "\\"O'Reilly Media, Inc.\\"",
    "title": "Scripting InDesign with JavaScript",
  },
  Object {
    "authors": Array [
      "Axel Rauschmayer",
    ],
    "coverUrl": "http://books.google.com/books/content?id=qU3rAgAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "Like it or not, JavaScript is everywhere these days—from browser to server to mobile—and now you, too, need to learn the language or dive deeper than you have. This concise book guides you into and through JavaScript, written by a veteran programmer who once found himself in the same position. Speaking JavaScript helps you approach the language with four standalone sections. First, a quick-start guide teaches you just enough of the language to help you be productive right away. More experienced JavaScript programmers will find a complete and easy-to-read reference that covers each language feature in depth. Complete contents include: JavaScript quick start: Familiar with object-oriented programming? This part helps you learn JavaScript quickly and properly. JavaScript in depth: Learn details of ECMAScript 5, from syntax, variables, functions, and object-oriented programming to regular expressions and JSON with lots of examples. Pick a topic and jump in. Background: Understand JavaScript’s history and its relationship with other programming languages. Tips, tools, and libraries: Survey existing style guides, best practices, advanced techniques, module systems, package managers, build tools, and learning resources.",
    "id": "qU3rAgAAQBAJ",
    "publishedDate": 2014-02-25T00:00:00.000Z,
    "publisher": "\\"O'Reilly Media, Inc.\\"",
    "title": "Speaking JavaScript",
  },
  Object {
    "authors": Array [
      "David Sawyer McFarland",
    ],
    "coverUrl": "http://books.google.com/books/content?id=CxY3lNQTATQC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
    "description": "JavaScript is an essential language for creating modern, interactive websites, but its complex rules challenge even the most experienced web designers. With JavaScript: The Missing Manual, you'll quickly learn how to use JavaScript in sophisticated ways -- without pain or frustration -- even if you have little or no programming experience. JavaScript expert David McFarland first teaches you the basics by having you build a simple program. Then you'll learn how to work with jQuery, a popular library of pre-built JavaScript components that's free and easy to use. With jQuery, you can quickly build modern, interactive web pages -- without having to script everything from scratch! Learn how to add scripts to a web page, store and manipulate information, communicate with the browser window, respond to events like mouse clicks and form submissions, and identify and modify HTML Get real-world examples of JavaScript in action Learn to build pop-up navigation bars, enhance HTML tables, create an interactive photo gallery, and make web forms more usable Create interesting user interfaces with tabbed panels, accordion panels, and pop-up dialog boxes Learn to avoid the ten most common errors new programmers make, and how to find and fix bugs Use JavaScript with Ajax to communicate with a server so that your web pages can receive information without having to reload",
    "id": "CxY3lNQTATQC",
    "publishedDate": 2008-07-21T00:00:00.000Z,
    "publisher": "\\"O'Reilly Media, Inc.\\"",
    "title": "JavaScript: The Missing Manual",
  },
]
`;
